kasa/discover.py

```diff
@@ -378,20 +380,30 @@ def _get_device_instance(
                 f"Unable to read response from device: {ip}: {ex}"
             ) from ex
 
-        if (
-            discovery_result.mgt_encrypt_schm.encrypt_type == "KLAP"
-            and discovery_result.mgt_encrypt_schm.lv is None
-        ):
+        if discovery_result.mgt_encrypt_schm.encrypt_type in ("KLAP", "AES"):
             type_ = discovery_result.device_type
             device_class = None
-            if type_.upper() == "IOT.SMARTPLUGSWITCH":
-                device_class = SmartPlug
+            supported_device_types: dict[str, Type[SmartDevice]] = {
+                "SMART.TAPOPLUG": TapoPlug,
+                "SMART.KASAPLUG": TapoPlug,
+                "IOT.SMARTPLUGSWITCH": SmartPlug,
+            }
+
+            if discovery_result.device_type in supported_device_types:
+                device_class = supported_device_types[discovery_result.device_type]
 
             if device_class:
                 _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
                 device = device_class(ip, port=port, credentials=credentials)
                 device.update_from_discover_info(discovery_result.get_dict())
-                device.protocol = TPLinkKlap(ip, credentials=credentials)
+                if discovery_result.mgt_encrypt_schm.encrypt_type == "KLAP":
+                    device.protocol = TPLinkKlap(ip, credentials=credentials)
+                else:
+                    device.protocol = TPLinkAes(
+                        ip,
+                        credentials=credentials,
+                        login_version=discovery_result.mgt_encrypt_schm.lv,
```

Comment:
I'm wondering if we should already pass the version even when it's not used? This would also allow storing the class ref in the map and pass the same parameters for all supported protocols to handle as they wish, something like this:
```
device.protocol = klass(ip=ip, credentials=credentials, encryption_schema=discovery_result.mgt_encrypt_schm)
```

Reply:
The `login_version` doesn't seem to do what I thought it would when I first added it so I have removed it for now.  I think we should move the discussion of what to pass to the protocol to https://github.com/python-kasa/python-kasa/pull/557 because on this PR it's not needed.